dali/operators/reader/loader/video/frames_decoder.cc

```diff
@@ -365,33 +365,65 @@ bool FramesDecoder::IsFormatSeekable() {
 }
 
 void FramesDecoder::BuildIndex() {
-  // TODO(awolant): Optimize this function for:
-  //  - CFR
-  //  - index present in the header
-
   index_ = vector<IndexEntry>();
 
   int last_keyframe = -1;
-  while (ReadRegularFrame(nullptr, false)) {
+  while (av_read_frame(av_state_->ctx_, av_state_->packet_) >= 0) {
     IndexEntry entry;
-    entry.is_keyframe = av_state_->frame_->flags & AV_FRAME_FLAG_KEY;
-    entry.pts = av_state_->frame_->pts;
-    entry.is_flush_frame = false;
+    // We want to make sure that we call av_packet_unref in every iteration
+    auto packet = AVPacketScope(av_state_->packet_, av_packet_unref);
 
+    if (packet->stream_index != av_state_->stream_id_) {
+      continue;
+    }
+    if (packet->flags & AV_PKT_FLAG_KEY) {
+      if (index_->size() == 0) {
+        entry.is_keyframe = true;
+      }
+      auto codec_id = av_state_->ctx_->streams[packet->stream_index]->codecpar->codec_id;
+      if (codec_id == AV_CODEC_ID_H264) {
+        if (packet->size >= 5) {  // Ensure there's enough data to inspect
+          // NAL unit type is in the fifth byte for H.264
+          uint8_t nal_unit_type = packet->data[4] & 0x1F;
+          if (nal_unit_type == 5) {  // IDR frame for H.264
+            entry.is_keyframe = true;
+          }
+        }
+      } else if (codec_id == AV_CODEC_ID_HEVC) {
+      // Further check if it's an IDR frame for H.265
+        if (packet->size >= 6) {  // Ensure there's enough data to inspect
+          // NAL unit type is in the sixth byte for H.265
+          uint8_t nal_unit_type = (packet->data[4] >> 1) & 0x3F;
```

Comment:
The comment suggests that it should be:

```suggestion
          uint8_t nal_unit_type = (packet->data[5] >> 1) & 0x3F;
```

Reply:
Good spot, I need to check if the source I get this from has a misleading comment or the code.

Reply:
Fixed